### DISCUSSION TITLE
KeyPoint and landmarks

### DISCUSSION THREAD

**Discussion by dubing-bit** [stella-cv/stella_vslam#573] [Q&A] [answered]
I would like to get the keypoint with its for landmarks when I use the and `m_keypoints = m_frame_publisher->get_keypoints(); ``vector<std::shared_ptr<stella_vslam::data::landmark>> landmarks = m_last_keyframe->get_landmarks();` The number of keypoints and landmarks obtained are as follows:
```
(16148, 2)
(3132, 3)
```
What can I do to get a one-to-one correspondence between keypoints and landmarks?

**Comment by zicklag** [comment]
Yeah, I'm trying to import the `map.msg` file into COLMAP, but I can't find anything that ties the world-space landmarks to the image-space key-points.

**Comment by ymd-stella** [comment, marked as answer]
If what you want is a undistorted key point, you can achieve the following

```cpp
vector<std::shared_ptr<stella_vslam::data::landmark>> landmarks = m_last_keyframe->get_landmarks();
for (unsigned int idx = 0; idx < landmarks.size(); ++idx) {
    if (landmarks.at(idx)) { 
        auto undist_keypt = m_last_keyframe->frm_obs_.undist_keypts_.at(idx);
    }
}
```

Landmarks (that are not nullptr) are mapped to keypoints by index.

If you want keypoints that are not undistorted, there is no way to do that now.

**Reply by dubing-bit** [reply]
Thanks, that successfully solved my problem.
